tools/dotnet-linker/BackingFieldDelayHandler.cs

```diff
@@ -0,0 +1,101 @@
+using Mono.Cecil.Cil;
+using Mono.Linker;
+using Mono.Linker.Steps;
+using Xamarin.Tuner;
+
+namespace Xamarin.Linker {
+
+	// Problems:
+	// * `Dispose` set the generated backing fields to `null` which means
+	//   the linker will mark every backing fields, even if not used 
+	//   elsewhere (generally properties) inside the class
+	// * Backing fields increase the memory footprint of the managed peer
+	//   instance (for the type and all it's subclasses)
+	// * Backing fields also increase the app size. Not a huge problem as
+	//   they are all declared _weakly_ as `NSObject` but still...
+	//
+	// Solution:
+	// * When the linker process a `Dispose` method of an `NSObject` 
+	//   subclass with the _optimizable_ attribute then we remove the
+	//   method body. This way the linker cannot mark the fields.
+	// * Before saving back the assemblies we replace the cached method
+	//   body and NOP every field that were not marked by something else
+	//   than the `Dispose` method.
+	public class BackingFieldDelayHandler : ConfigurationAwareMarkHandler {
+
+		protected override string Name { get; } = "Backing Fields Optimizer";
+		protected override int ErrorCode { get; } = 2200;
+
+		public override void Initialize (LinkContext context, MarkContext markContext)
+		{
+			base.Initialize (context);
+			markContext.RegisterMarkMethodAction (ProcessMethod);
+		}
+
+		// cache `Dispose` body of optimization NSObject subclasses
+		static Dictionary<MethodDefinition, MethodBody> dispose = new ();
+		
+		protected override void Process (MethodDefinition method)
+		{
+			if (!method.HasParameters || !method.IsVirtual || !method.HasBody)
+				return;
+			if (method.Name != "Dispose")
+				return;
+			// only process methods that are marked as optimizable
+			if (!method.IsBindingImplOptimizableCode (LinkContext))
+				return;
+			var t = method.DeclaringType;
+			if (!t.IsNSObject (LinkContext))
+				return;
+
+			// keep original for later (if needed)
+			dispose.Add (method, method.Body);
+
+			// setting body to null will only cause it to be reloaded again
+			// same if we don't get a new IL processor
+			// and we do not want that (as it would mark the fields)
+			var body = new MethodBody (method);
+			var il = body.GetILProcessor ();
+			il.Emit (OpCodes.Ret);
+			method.Body = body;
+		}
+
+		public static void ReapplyDisposedFields (DerivedLinkContext context)
+		{
+			foreach (var kvp in dispose) {
+				var method = kvp.Key;
+				// note: all methods in the dictionary are marked
+				var body = kvp.Value;
+				foreach (var ins in body.Instructions) {
+					switch (ins.OpCode.OperandType) {
+					// case OperandType.InlineTok:
+					case OperandType.InlineField:
+						var field = (ins.Operand as FieldReference)?.Resolve ();
+						if (!context.Annotations.IsMarked (field)) {
+							ins.OpCode = OpCodes.Nop; // stfld
+							ins.Previous.OpCode = OpCodes.Nop; // ldnull
+							ins.Previous.Previous.OpCode = OpCodes.Nop; // ldarg.0
```

Comment:
You're assuming that the instructions are `ldarg.0 + ldnull + stdfld` here (which sounds like a reasonable assumption today), but that might not be true in the future, where a new C# compiler might end up with a different sequence. Maybe verify this assumption, and if not show an error/warning?

```csharp
if (ins == stdfld && ins.Previous == ldnull && ins.Previous.Previous == ldarg.0) {
    // yay!
} else {
    // nay?
}
```